openjdk/mmtkBarrierSetC1.hpp

```diff
@@ -4,27 +4,84 @@
 #include "c1/c1_CodeStubs.hpp"
 #include "gc/shared/c1/barrierSetC1.hpp"
 
+class MMTkBarrierSetAssembler;
+
 class MMTkBarrierSetC1 : public BarrierSetC1 {
-public:
-  MMTkBarrierSetC1() {}
+  friend class MMTkBarrierSetAssembler;
+
+protected:
+  CodeBlob* _write_barrier_c1_runtime_code_blob;
+
+  /// Full pre-barrier
+  virtual void object_reference_write_pre(LIRAccess& access, LIR_Opr src, LIR_Opr slot, LIR_Opr new_val) const {}
+  /// Full post-barrier
+  virtual void object_reference_write_post(LIRAccess& access, LIR_Opr src, LIR_Opr slot, LIR_Opr new_val) const {}
 
-  virtual void store_at_resolved(LIRAccess& access, LIR_Opr value) {
+  /// Substituting write barrier
+  virtual void store_at_resolved(LIRAccess& access, LIR_Opr value) override {
+    if (access.is_oop()) object_reference_write_pre(access, access.base().opr(), access.resolved_addr(), value);
     BarrierSetC1::store_at_resolved(access, value);
+    if (access.is_oop()) object_reference_write_post(access, access.base().opr(), access.resolved_addr(), value);
   }
 
-  virtual LIR_Opr atomic_cmpxchg_at_resolved(LIRAccess& access, LIRItem& cmp_value, LIRItem& new_value) {
-    return BarrierSetC1::atomic_cmpxchg_at_resolved(access, cmp_value, new_value);
+  /// Substituting write barrier (cmpxchg)
+  virtual LIR_Opr atomic_cmpxchg_at_resolved(LIRAccess& access, LIRItem& cmp_value, LIRItem& new_value) override {
+    if (access.is_oop()) object_reference_write_pre(access, access.base().opr(), access.resolved_addr(), new_value.result());
+    LIR_Opr result = BarrierSetC1::atomic_cmpxchg_at_resolved(access, cmp_value, new_value);
```

Comment:
Cmpxchg, when successful, performs both a read and a write operation.  But if cmpxchg fails, the write operation is not performed.  Should we skip the post write barrier if it fails?

What does cmpxchg do in vanilla OpenJDK?

Reply:
I was following OpenJDK's style, which does not check the result status either. Here're two examples:
* [C1 atomic xchg barrier](https://github.com/mmtk/openjdk/blob/master/src/hotspot/share/gc/shared/c1/modRefBarrierSetC1.cpp#L75-L81)
* [C2 atomic xchg barrier](https://github.com/mmtk/openjdk/blob/master/src/hotspot/share/gc/shared/c2/modRefBarrierSetC2.cpp#L73-L81)

Reply:
I think it is OK to apply card marking and object remembering barriers conservatively even when the object is not actually changed.  It is OK for now, but I am not sure if other barriers are OK, too.  Let's discuss this separately.